pkg/reconciler/taskrun/taskrun.go

```diff
@@ -103,11 +103,16 @@ func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
 
 	// If the TaskRun is just starting, this will also set the starttime,
 	// from which the timeout will immediately begin counting down.
-	tr.Status.InitializeConditions()
-	// In case node time was not synchronized, when controller has been scheduled to other nodes.
-	if tr.Status.StartTime.Sub(tr.CreationTimestamp.Time) < 0 {
-		c.Logger.Warnf("TaskRun %s createTimestamp %s is after the taskRun started %s", tr.GetRunKey(), tr.CreationTimestamp, tr.Status.StartTime)
-		tr.Status.StartTime = &tr.CreationTimestamp
+	if !tr.HasStarted() {
+		tr.Status.InitializeConditions()
+		// In case node time was not synchronized, when controller has been scheduled to other nodes.
+		if tr.Status.StartTime.Sub(tr.CreationTimestamp.Time) < 0 {
+			c.Logger.Warnf("TaskRun %s createTimestamp %s is after the taskRun started %s", tr.GetRunKey(), tr.CreationTimestamp, tr.Status.StartTime)
+			tr.Status.StartTime = &tr.CreationTimestamp
+		}
+		// Emit events
+		afterCondition := tr.Status.GetCondition(apis.ConditionSucceeded)
+		reconciler.EmitEvent(c.Recorder, nil, afterCondition, tr)
```

Comment:
i'm a bit confused why this call to `EmitEvent` is needed in addition to the other one a bunch of lines down in the same function

is it possible we can get away with one? i do like the idea of only having to call this logic once instead of being aware of 2 places it is called

or if it is required maybe a comment explaining why?

Reply:
You need to emit the event every time the condition changes, because - as things are today - you could have more than one condition change in one reconcile cycle.

That could change if we added extra conditions in future, i.e. instead of only using Succeeded (unknown/true/false) we could use Started and Running too. In any case, I think it's worth sending the event as early as possible. This reconcile cycle will go on to create a pod and run several API calls, so we may be losing a few hundreds milliseconds if we wait. 

If something is looking for the event to update the status of a check in github, half a second later makes a difference.

Reply:
>  you could have more than one condition change in one reconcile cycle.

sounds like that's only in the started case? going from started to running - and shouldn't be expected from any other states? 

anyway makes sense! i do think this is worth a comment explaining why the 2 calls but also wouldnt block on this

Reply:
Yeah, it's only in that case. I promise I'll add a comment in a follow-up 👼